### ISSUE TITLE
ActionText does not support Trix content attachments

### ISSUE THREAD

**Issue by micahbf** [rails/rails#35218]
### Steps to reproduce
1. Create a model that `has_rich_text` and a form for it with a `rich_text_area`
2. Load the form, and in the browser JS console run:
```js
attachment = new Trix.Attachment({content: '<iframe src="https://player.vimeo.com/video/70591644?title=0&byline=0&portrait=0" width="640" height="360" frameborder="0" webkitallowfullscreen mozallowfullscreen allowfullscreen></iframe>'})
editor = document.querySelector('trix-editor').editor
editor.insertAttachment(attachment)
```
3. Observe that the Trix editor displays the embedded attachment, and that the hidden form input's value contains the HTML for the attachment.
4. Submit the form

### Expected behavior
The Trix content attachment is saved with the rest of the rich content

### Actual behavior
The content attachment is dropped and not persisted

### System configuration
**Rails version**: 6.0.0.beta1

**Ruby version**: 2.6.1

**Comment by matthee** [comment]
@micahbf Are you sure, the attachment isn't persisted to the DB? I've run into a similar issue and tracked it down to the HTML sanitizer removing non-whitelisted tags. 

I've added the missing tag to the whitelist using the following initializer. Bear in mind that adding `iframe` to the whitelisted tags might be a security risk. I suppose you might also want to customize the `allowed_attributes` field of the `WhitelistSanitizer`.

```ruby
# config/initializers/html_sanitation.rb

Rails::Html::WhiteListSanitizer.allowed_tags << "tagname"
``` 

See: https://github.com/rails/rails-html-sanitizer/blob/f82bfd25c02f806e51b5e2ae62334bba4e908ad8/lib/rails/html/sanitizer.rb#L107

**Comment by micahbf** [comment]
@matthee Interesting, that hadn't occurred to me. I just checked and the attachment is not persisted to the DB. But, I'll play around with this more later and see what else I can find about when exactly this gets stripped out.

**Comment by excid3** [comment]
This isn't documented anywhere yet on how ActionText stores attachments, but here's a quick explanation. I'm planning on writing up some more stuff for the docs.

ActionText requires attachments to have a signed global ID (sgid). It minifies the HTML you insert on save, so that it can re-render it with up-to-date content later on. This makes it so you can reference models and always display the current content when those records change. 

The `sgid` comes from the ActionText::Attachable module's `attachable_sgid` method.

You have to set up something like this:

```ruby
class Video < ApplicationController
  include ActionText::Attachable
end
```

This gives your videos `video.attachable_sgid` which you'll have to send over to the client for inserting with your `Trix.Attachment`.

```js
attachment = new Trix.Attachment({
  sgid: video.attachable_sgid,
  content: '<iframe src="https://player.vimeo.com/video/70591644?title=0&byline=0&portrait=0" width="640" height="360" frameborder="0" webkitallowfullscreen mozallowfullscreen allowfullscreen></iframe>'
})
```

ActionText will load up the model from the global id and then render it with the default partial path when you render the content. In this case it would be `videos/_video.html.erb`.

Unfortunately, I don't think there are any good ways of inserting attachments right now that aren't backed by a GlobalID referenceable model.

**Comment by micahbf** [comment]
Thank you @excid3, that is super helpful.

**Comment by matthee** [comment]
Thanks @excid3, that's a great insight into ActionText internals! I assume you meant to derive `Video` from `ApplicationRecord` instead from `ApplicationController`?

@micahbf I was just having a look at my code once again. The proposed fix only seems to work with custom tags, but not with attachments. My bad.

**Comment by excid3** [comment]
@matthee hah! Yes, fixed my code example.

**Comment by inopinatus** [comment]
Um. So all my attachments will break when we rotate `secret_key_base` ?!?

**Comment by travisp** [comment]
If attachments must have an sgid, does this mean that the proposed solution to allow HR tags in trix can't work with actiontext? https://github.com/basecamp/trix/issues/570 (inserting as a content attachment)?

Horizontal rules may be a bit of an odd case, but it would be really nice to be able to support them.

**Comment by codergeek121** [comment]
> If attachments must have an sgid, does this mean that the proposed solution to allow HR tags in trix can't work with actiontext? [basecamp/trix#570](https://github.com/basecamp/trix/issues/570) (inserting as a content attachment)?
> 
> Horizontal rules may be a bit of an odd case, but it would be really nice to be able to support them.

### `<hr>` tags

The `<hr>` tag with ActionText is already working.
1. We need to insert a `Trix.Attachment`. It needs to have the `contentType` set to `vnd.rubyonrails.horizontal-rule.html`
```js
const trixEditorTag = document.querySelector("trix-editor");
const attachment = new Trix.Attachment({
    content: "<hr>",
    contentType: "vnd.rubyonrails.horizontal-rule.html"
  })
  trixEditorTag.editor.insertAttachment(attachment);
```
Here's the reason why: https://github.com/rails/rails/blob/67705c8bbfa6f6d0e3be1208741a947143111786/actiontext/lib/action_text/attachables/content_attachment.rb#L8-L18
2. We need to define two partials.
The first is to render the `<hr>` tag in `<trix-editor>` when loading the content from the database. 
```
views/action_text/attachables/content_attachments/_horizontal_rule.html.erb
<hr>
```
The second partial gets rendered when displaying the rich content.
```
views/action_text/attachables/_content_attachment.html.erb
<hr>
```

Is this something we should write documentation for? (#34978)
It could also make sense to update Trix to include a `<hr>` toolbar-button and add these partials to ActionText so it's possible to use all features of ActionText? I'd happily help to improve the documentation or code :)

### Custom content Attachments
Since this solution is based on the already implemented `Attachables::ContentAttachment` it's possible to use `<hr>` tags without a database model. As far as I understand to solve this issue we would need a way to build `HTMLAttachments` like `ContentAttachment` but configurable by the developer? Is this something ActionText tries to solve, or is this out of scope and too narrow of a use case?

**Comment by sosodev** [comment]
@codergeek121 I'm not sure if it was in scope of the original project, but I would love to see ActionText be configurable in such a rails-esque way.

**Comment by sosodev** [comment]
Just to summarize what's been said here so far

Currently ActionText supports 2 (3ish) kinds of attachments [[0]](https://github.com/rails/rails/blob/master/actiontext/lib/action_text/attachable.rb#L11)

- attachments that are linked to a `sgid` [[1]](https://github.com/rails/rails/issues/35218#issuecomment-470586641)
- image attachments [[2]](https://github.com/rails/rails/blob/master/actiontext/lib/action_text/attachables/remote_image.rb)

Missing is support for static HTML attachments. Currently the implementation for those is half-finished [[3]](https://github.com/rails/rails/blob/master/actiontext/lib/action_text/attachables/content_attachment.rb) and only supports horizontal rules [[4]](https://github.com/rails/rails/issues/35218#issuecomment-512825833).

All three types are expected to render out of partials, so without refactoring support for static Trix attachments probably instead means support for attaching static partials.

I think a naive solution could simply add a new content-type to [ContentAttachment](https://github.com/rails/rails/blob/master/actiontext/lib/action_text/attachables/content_attachment.rb) that specifies the attachment as static and then renders the matching name from `action_text/attachables/content_attachments/#{name.underscore}`

**Comment by ajessee** [comment]
Hi there. New Rails dev, trying to hack my way through things. 

I'm trying to embed github gists into my blog post using trix and action text. I've reviewed the info above, and my understanding is that I have two options:

1. Wait until support for static HTML attachments is built into ContentAttachment. Or
2. Fork the rails gem, and update ContentAttachment myself.

> I think a naive solution could simply add a new content-type to ContentAttachment that specifies the attachment as static and then renders the matching name from action_text/attachables/content_attachments/#{name.underscore}

Can anyone provide an example of how to implement this 'naive' solution? I'm at a loss.

**Comment by sosodev** [comment]
@ajessee Hey, I actually [did](https://github.com/sosodev/actiontext-static) create the naive solution and I've been using it fine in production on an [MVP](https://tryfosh.com) I created (lol). If you're having trouble getting it to work feel free to create an issue on the repo.

**Comment by ajessee** [comment]
@sosodev wow! Thanks so much for your time and response. I'll look into your gem and see if I can implement it.

My use case is to try to embed a github gist as static html. What kind of breaking changes in the validation should I look out for?

**Comment by sosodev** [comment]
@ajessee Should be fine for your use case, I just mean it's dangerous if you're letting users write.

**Comment by ajessee** [comment]
Hey @sosodev - I'm getting back to this now, and I know in your readme for the gem, it says to patch your config using `ActionText::Static.patch(config)`.

Where would I do this? At first I thought I could put it in an initializer or in config/application.rb, but when I do, I'm getting a `uninitialized constant ActionText::Static (NameError)` when I fire up rails server. I've tried to require 'actiontext/static', but that doesn't work either.

I appreciate the help, sorry for the barrage of questions!

**Comment by ajessee** [comment]
Hi @sosodev - Sorry to bother, but I can't seem to figure this out. Is there a rake task that I need to run to patch the rails code with the new static module?

**Comment by sosodev** [comment]
Hi, @ajessee I created an issue on the gem's repository with a better explanation and tagged you in it so we don't have to clutter this issue. Not totally sure how github notifications work but I thought that would ping you. Sorry for the confusion.

**Comment by lazaronixon** [comment]
https://gist.github.com/lazaronixon/5bd92873a4b7d21ebf4ce6e0ff930a62
Trix with h1/h2/h3/h4/h5/h6/hr

**Comment by lazaronixon** [comment]
https://github.com/lazaronixon/trix-extensions
![trixext](https://user-images.githubusercontent.com/2651240/72076801-e97d9b00-32d4-11ea-845e-671ba999f560.gif)

**Comment by matthee** [comment]
@lazaronixon Great additions! 

I would love to see these features being merged into ActionText.

**Comment by healerjoy** [comment]
@excid3 Thanks for explaining the method to implement this using ActionText::Attachable. I was able to get this working with default partial path.

Is there a solution to use a custom partial path?
I found a method that can be overridden called [to_trix_content_attachment_partial_path](https://github.com/rails/rails/blob/28775ff3a231253c0052531f7d4e6b264b15497b/actiontext/lib/action_text/attachable.rb#L70). ActionText is using this in TrixConversion to allow for the custom partial path. Is this the right method to override?

https://github.com/rails/rails/blob/28775ff3a231253c0052531f7d4e6b264b15497b/actiontext/lib/action_text/attachments/trix_conversion.rb#L30

After some investigation I found that ContentHelper does not use this method to allow for the custom partial path.
https://github.com/rails/rails/blob/28775ff3a231253c0052531f7d4e6b264b15497b/actiontext/app/helpers/action_text/content_helper.rb#L24
https://github.com/rails/rails/blob/28775ff3a231253c0052531f7d4e6b264b15497b/actiontext/app/helpers/action_text/content_helper.rb#L30

I am still new to learning the internals of ActionText. Hope the above information is helpful in answering the question.

**Comment by excid3** [comment]
ActionText renders `to_trix_content_attachment_partial_path` for editing in Trix and `to_partial_path` when rendering in the ContentHelper @healerjoy. You can override either method to change how those work.

Originally, `to_trix_content_attachment_partial_path` was undefined and you had to implement it yourself, but my PR was merged to default that to `to_partial_path` so it would be the same in editing and rendering.

Sometimes you need these to different partials, like if you're embedding videos, you'll want a thumbnail in Trix, but the actual iframe in the rendered version.

My Railsconf Couch Edition talks about all this in-depth with examples on YouTube embeds and user mentions, so be on the lookout for that!

**Comment by healerjoy** [comment]
@excid3 Ah! This is great insight. Thanks for your reply. Now I understand that the true purpose of `to_trix_content_attachment_partial_path` method is to render only when editing with Trix Editor.

**Use Case**
I use a model called `Person` which renders the default partial `person/_person.html.erb` using the default `to_partial_path` method. This partial is already being used to render data of the person.

When I use `ActionText::Attachable` on this model and override the `to_trix_content_attachment_partial_path` method with a custom partial path, then the editing in Trix Editor works well.

However, the view content is still loaded using the ContentHelper which uses the `to_partial_path` method. As it can be noticed, it is not possible to override `to_partial_path` in this case because it is already being used for another purpose.

I am using the above method to build @mentions in Trix Editor using the `Person` model without the need to have a separate `Mention` model.

Do you feel this is a bug/needed-improvement to allow for a custom partial path? Or the other understanding can be that `ActionText::Attachable` should only be used on a separate model that will work only as attachable. Currently, the separate model approach works properly because the `to_partial_path` of the separate model can be used as needed.

Looking forward to the **Railsconf Couch Edition** talk.

**Comment by excid3** [comment]
I think it would be nice if there was a separate `to_content_partial_path` that it used so you could override it separate from the `to_partial_path` that you might be using for other things. Right now you would have to override the `ContentHelper` method to change the `render` call.

**Comment by healerjoy** [comment]
Great. That makes sense and will make the ActionText partial rendering independent of `to_partial_path`. Do you feel this can be a positive addition to ActionText and is inline with the overall direction of the framework? I would be happy to work on it and create a pull request for `to_content_partial_path`.

**Comment by excid3** [comment]
It sounds like they've made some extensive changes to ActionText that aren't out yet for Hey, so it's probably a good idea to wait on that and see what's changed. I assume that'll get merged in Rails 6.1.

**Comment by healerjoy** [comment]
Ok, sounds great. Thank you for your excellent insights and this conversation @excid3. Let's await the new changes to ActionText.

**Comment by tbcooney** [comment]
It's a shame when the bot marks the issue stale when it's still a relevant issue 😞

**Comment by marckohlbrugge** [comment]
@healerjoy @excid3 If you don't want to override `ContentHelper`'s `render_action_text_attachments` method you can instead check for `local_assigns.key?(:in_gallery)` in your partial. It seems like this is always assigned when rendering through Action Text.

It's definitely still a hack, but doesn't require monkey patching and is easy to undo if Rails 6.1 does indeed add support for dedicated partials. 

```erb
<% if local_assigns.key?(:in_gallery) %>
  <!-- render what an @mention in action text should look like -->
<% else %>
  <!-- your existing people/person partial code -->
<% end %>
```

**Comment by excid3** [comment]
@marckohlbrugge slick! That's great to know. I feel like ideally this changes in a future version, but at least that's not _too_ bad for now.

**Comment by healerjoy** [comment]
@marckohlbrugge thanks for this useful information. Will keep this in mind when revisiting that part of code.

For anyone's reference, when I was working on this problem, I proceeded with the approach of using a separate database-less ruby model to support my use case.

**Comment by seanharmer** [comment]
I just hit this exact issue too with embedding wistia-hosted videos into ActionText. I wanted to offer two ways to do this, one with a "standard" inline video player and another with a small thumbnail which when clicked launches a popup modal video player.

I too took the same approach as @healerjoy and used a non-db model for the popup videos.

It feels like the model is the wrong place to customise the appearance of how a resource should be rendered as an attachment. Feels more like it belongs on a controller-like object.

**Comment by lylo** [comment]
> Um. So all my attachments will break when we rotate `secret_key_base` ?!?

Does anyone have any thoughts or feelings about this? Feels like a potential flaw in this approach?

**Comment by Maxhou00** [comment]
Every attachment are lost and replaced by a ☒ when `credentials.ymc` is changed.

**Comment by tbcooney** [comment]
> @healerjoy @excid3 If you don't want to override `ContentHelper`'s `render_action_text_attachments` method you can instead check for `local_assigns.key?(:in_gallery)` in your partial. It seems like this is always assigned when rendering through Action Text.
> 
> It's definitely still a hack, but doesn't require monkey patching and is easy to undo if Rails 6.1 does indeed add support for dedicated partials.
> 
> ```
> <% if local_assigns.key?(:in_gallery) %>
>   
> <% else %>
>   
> <% end %>
> ```

Is this still one of the "preferred" methods, or has there been a release that provides a better experience for this?

**Comment by aantix** [comment]
This thread was a huge help.  

I had a div with an inline video that I wanted to add as an attachment to the Trix document (drag n drop).

For the partial:

```
# idea_items/_idea_item_video.html.erb
<div id="<%= dom_id(idea_item_video) %>"
     ....
     data-sgid="<%= idea_item_video.attachable_sgid %>"
     ....
```

For the javascript:

```
      const attachment = new Trix.Attachment({content: clone.outerHTML,
                                             sgid: draggedItem.dataset.sgid});  // <--- sgid of the video div
      editor.insertAttachment(attachment)
```

For the model, since the partial was in a non-default directory, I had to override `to_partial_path`

```
class IdeaItemVideo < IdeaItem
  def to_partial_path
    'idea_items/idea_item_video'
  end
```
